linera-views/src/reentrant_collection_view.rs

```diff
@@ -889,25 +933,27 @@ where
         let mut hasher = sha3::Sha3_256::default();
         let keys = self.keys().await?;
         hasher.update_with_bcs_bytes(&keys.len())?;
-        let updates = self.updates.get_mut();
+        let cached_entries = self.cached_entries.get_mut();
         for key in keys {
             hasher.update_with_bytes(&key)?;
-            let hash = match updates.get_mut(&key) {
-                Some(entry) => {
-                    let Update::Set(view) = entry else {
-                        unreachable!();
-                    };
-                    let mut view = view
-                        .try_write_arc()
-                        .ok_or_else(|| ViewError::TryLockError(key))?;
-                    view.hash_mut().await?
-                }
-                None => {
-                    let key = self.context.base_tag_index(KeyTag::Subview as u8, &key);
-                    let context = self.context.clone_with_base_key(key);
-                    let mut view = W::load(context).await?;
-                    view.hash_mut().await?
-                }
+            let hash = if let Some(entry) = self.updates.get_mut(&key) {
+                let Update::Set(view) = entry else {
+                    unreachable!();
+                };
+                let mut view = view
+                    .try_write_arc()
+                    .ok_or_else(|| ViewError::TryLockError(key))?;
+                view.hash_mut().await?
+            } else if let Some(view) = cached_entries.get_mut(&key) {
+                let mut view = view
+                    .try_write_arc()
+                    .ok_or_else(|| ViewError::TryLockError(key))?;
+                view.hash_mut().await?
```

Comment:
Does `cached_entries` still reflect storage after this? Won't the hash be only in the view in memory, but not in storage?

Reply:
It does, because `hash_mut` shouldn't (in theory) change the underlying view. It's just an optimization of `hash` that avoids locking because it has a `&mut self` receiver and knows any locks can be immediately acquired (except entry-specific locks in `ReentrantCollectionView`, IIRC).

Regarding keeping the hash in memory, I'm not sure I fully understand, but if you mean the `WrappedHashableContainerView`, then both `hash` and `hash_mut` update the hash stored in memory (and I think that's okay, as long as the hash only changes when the underlying view changes).